clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -276,27 +329,83 @@ else if (expireMs <= now) {
+        clusterResourceListeners.onUpdate(clusterForListeners.clusterResource());
 
         notifyAll();
-        log.debug("Updated cluster metadata version {} to {}", this.version, this.cluster);
+        log.debug("Updated cluster metadata version {} to {}", this.version, this.cache);
+    }
+
+    /**
+     * Transform a MetadataResponse into a new MetadataCache instance.
+     * @param metadataResponse
+     * @param topicsToRetain
+     * @return
+     */
+    private MetadataCache handleMetadataResponse(MetadataResponse metadataResponse, Predicate<String> topicsToRetain) {
+        Set<String> internalTopics = new HashSet<>();
+        List<PartitionInfo> partitions = new ArrayList<>();
+        for (MetadataResponse.TopicMetadata metadata : metadataResponse.topicMetadata()) {
+            if (!topicsToRetain.test(metadata.topic()))
+                continue;
+
+            if (metadata.error() == Errors.NONE) {
+                if (metadata.isInternal())
+                    internalTopics.add(metadata.topic());
+                for (MetadataResponse.PartitionMetadata partitionMetadata : metadata.partitionMetadata()) {
+                    updatePartitionInfo(metadata.topic(), partitionMetadata, partitions::add);
+                }
+            }
+        }
+
+        return new MetadataCache(metadataResponse.clusterId(), new ArrayList<>(metadataResponse.brokers()), partitions,
+                metadataResponse.topicsByError(Errors.TOPIC_AUTHORIZATION_FAILED),
+                metadataResponse.topicsByError(Errors.INVALID_TOPIC_EXCEPTION),
+                internalTopics, metadataResponse.controller());
+    }
+
+    /**
+     * Compute the correct PartitionInfo to cache for a topic+partition.
+     * @param topic
+     * @param partitionMetadata
+     * @param partitionInfoConsumer
+     */
+    private void updatePartitionInfo(String topic, MetadataResponse.PartitionMetadata partitionMetadata, Consumer<PartitionInfo> partitionInfoConsumer) {
+        TopicPartition tp = new TopicPartition(topic, partitionMetadata.partition());
+        if (partitionMetadata.leaderEpoch().isPresent()) {
+            int newEpoch = partitionMetadata.leaderEpoch().get();
+            if (updateLastSeenEpoch(tp, newEpoch, oldEpoch -> newEpoch >= oldEpoch) == newEpoch) {
+                // If the received leader epoch is at least the same as the previous one, use the new partition info
+                partitionInfoConsumer.accept(MetadataResponse.partitionMetaToInfo(topic, partitionMetadata));
+            } else {
+                // Otherwise ignore the new metadata and use the previously cached info
+                PartitionInfo previousInfo = cache.cluster().partition(tp);
+                if (previousInfo != null) {
+                    partitionInfoConsumer.accept(previousInfo);
+                } else {
+                    log.warn("Got an older epoch in partition metadata response for {}, but could not find previous partition " +
```

Comment:
This case might happen if the set of topics is updated with a Metadata request in flight? If that's a common case, perhaps we should reduce this to DEBUG?

Reply:
Yea good point -- if we "forget" about epochs from topics we're not tracking anymore we could hit this case. I'll add a check if the TopicPartition we got is in the `topics` set for this case